text/0097-log-data-model.md

```diff
@@ -0,0 +1,972 @@
+  of different types. The keys and values for well-known fields follow semantic
+  conventions for key names and possible values that allow all parties that work
+  with the field to have the same interpretation of the data. See references to
+  semantic conventions for `Resource` and `Attributes` fields and examples in
+  [Appendix A](#appendix-a-example-mappings).
+
+The reasons for having these 2 kinds of fields are:
+
+- Ability to efficiently represent named top-level fields, which are almost
+  always present (e.g. when using encodings like Protocol Buffers where fields
+  are enumerated but not named on the wire).
+  
+- Ability to enforce types of named fields, which is very useful for compiled
+  languages with type checks.
+
+- Flexibility to represent less frequent data via key/value pair lists. This
+  includes well-known data that has standardized semantics as well as arbitrary
+  custom data that the application may want to include in the logs.
+
+When designing this data model I followed the following reasoning to make a
+decision about when to use use a top-level named field:
+
+- The field needs to be either mandatory for all records or be frequently
+  present in well-known log and event formats (such as `Timestamp`) or is
+  expected to be often present in log records in upcoming logging systems (such
+  as `TraceId`).
+
+- The field’s semantics must be the same for all known log and event formats and
+  can be mapped directly and unambiguously to this data model.
+  
+Both of the above conditions were required to give the field a place in the
+top-level structure of the record.
+
+## Log and Event Record Definition
+
+Note: below we use type `any`, which can be a scalar value (number, string or
+boolean), or an array or map of values. Arbitrary deep nesting of values for
+arrays and maps is allowed (essentially allow to represent an equivalent of a
+JSON object).
+
+[Appendix A](#appendix-a-example-mappings) contains many examples that show how
+existing log formats map to the fields defined below. If there are questions
+about the meaning of the field reviewing the examples may be helpful.
+
+Here is the list of fields in a log record:
+
+Field Name     |Description
+---------------|--------------------------------------------
+Timestamp      |Time when the event occurred.
+TraceId        |Request trace id.
+SpanId         |Request span id.
+TraceFlags     |W3C trace flag.
+SeverityText   |The severity text (also known as log level).
+SeverityNumber |Numerical value of the severity.
+ShortName      |Short event identifier.
+Body           |The body of the log record.
+Resource       |Describes the source of the log.
+Attributes     |Additional information about the event.
+
+Below is the detailed description of each field.
+
+### Field: `Timestamp`
+
+Type: Timestamp, uint64 nanoseconds since Unix epoch.
+
+Description: Time when the event occurred measured by the origin clock. This
```

Comment:
I think this is ambiguous, because the time that the error occurred may be different on several different levels: it could mean the time that the business event occured (before it was logged), the time that the event was logged (entering the logging framework), or the time that the logs was serialized in a particular format (written to disk, memory, kafka topic).

Reply:
It is true that the time the event occurred may be different from the time it was log ged and different from the time it was serialized.

However, often in logging systems there important differentiation is between time of occurrence and time of arrival to backend (because some events lack the occurrence timestamp and are timestampted on arrival).

The finer differentiation between 3 different cases you described appears to be less common in the industry. If necessary it can be expressed using additional attributes (feel free to propose a convention).

The intent of this field is that this is the best approximation of the time when the event occurred. Please feel free to suggest a better wording.

Reply:
Resolving, please feel free to reopen if there are further thoughts.

Reply:
> The finer differentiation between 3 different cases you described appears to be less common in the industry. 

This isn't an abstract concern -- it happens in Logback consistently. 

Consider a situation where a duration is measured in a Java application, using a timer.  On conclusion of the operation, the duration is measured down to the nanosecond, and rendered:

```java
logger.info("Operation completed successfully, started {}, took {}", startTime, duration);
```

If you looked at the time of the logging and the start time, you'd expect the duration of logging and the duration of start time to be the same.  However, that's not what happens.  Logback creates a `LoggingEvent` [with the timestamp in constructor](https://github.com/qos-ch/logback/blob/master/logback-classic/src/main/java/ch/qos/logback/classic/spi/LoggingEvent.java#L117), and that only happens after evaluation of [turbo filters](http://logback.qos.ch/manual/filters.html#TurboFilter).  The result is a small but non-zero lag between when the event "occured" and the time a logging event was created for it.

Likewise, timestamp definitions are also not rhetorical.  [Honeycomb Tracing](https://docs.honeycomb.io/working-with-your-data/tracing/send-trace-data/#manual-tracing), for example, understands the event to have occurred at the start of the operation, and not at the end, and understands the end as "timestamp + duration_ms".

Reply:
Makes sense. Attributes is likely the best place to record additional timestamps for the use cases where the difference is important.